src/shared_utilities/FileReader.jl

```diff
@@ -1,549 +1,100 @@
-    regrid_dirpath::String,
-    get_infile::Function,
-    varnames::Vector{String},
-    date_ref::Union{DateTime, DateTimeNoLeap},
-    t_start,
-    surface_space::Spaces.AbstractSpace;
-    mono::Bool = true,
-) where {FT <: AbstractFloat}
-    comms_ctx = ClimaComms.context(surface_space)
-    outfile_root = "temporal_data_cgll"
-
-    # Initialize dummy date to be overwritten by actual dates during file read
-    all_dates = [DateTime(0)]
-
-    # Regrid data at all times from lat/lon (RLL) to simulation grid (CGLL)
-    # Download `infile_path` artifact on root process first to avoid race condition
-    if ClimaComms.iamroot(comms_ctx)
-        infile_path = get_infile()
-        all_dates = Regridder.hdwrite_regridfile_rll_to_cgll(
-            FT,
-            regrid_dirpath,
-            infile_path,
-            varnames,
-            surface_space,
-            outfile_root;
-            mono = mono,
-        )
-
-        NCDataset(infile_path, "r") do ds
-            if !("time" in keys(ds))
-                error(
-                    "Using a temporal albedo map requires data with time dimension.",
-                )
-            end
-        end
-    end
-    all_dates = ClimaComms.bcast(comms_ctx, all_dates)
-    ClimaComms.barrier(comms_ctx)
-    infile_path = get_infile()
-
-    # Init time tracking info
-    data_fields =
-        Fields.zeros(FT, surface_space), Fields.zeros(FT, surface_space)
-    # Store `segment_length` as an array so we can modify it as a field of a struct
-    segment_length = Int[0]
-
-    date_start = date_ref + Dates.Second(t_start)
-    if date_start < all_dates[1]
-        @warn "Simulation start date is before first file data"
-    end
+function NCFileReader(file_path::AbstractString, varname::AbstractString)
+    dataset = NCDatasets.NCDataset(file_path)
 
-    # Find the index of the start file closest to this simulation's start date
-    # Like `segment_length`, store in an array so we can modify in struct
-    date_idx0 =
-        [argmin(abs.(Dates.value(date_start) .- Dates.value.(all_dates[:])))]
+    # NOTE: We are hardcoding a few things here!
+    dimensions = dataset["lon"][:], dataset["lat"][:]
+    available_dates = read_available_dates(dataset)
 
-    # Construct component structs of PrescribedDataTemporal object
-    file_info = FileInfo(
-        infile_path,
-        regrid_dirpath,
-        varnames,
-        outfile_root,
-        all_dates,
-        date_idx0,
-    )
-    file_states = Dict{String, FileState{typeof(data_fields)}}()
-    for varname in varnames
-        file_states[varname] =
-            FileState(deepcopy(data_fields), copy(date_idx0), segment_length)
+    if !isempty(available_dates)
+        issorted(available_dates) ||
+            error("Cannot process files that are not sorted in time")
     end
-    sim_info = SimInfo(date_ref, t_start)
 
-    args = (file_info, file_states, sim_info)
+    _cached_reads = Dict{Dates.DateTime, Array}()
```

Comment:
in the future, this is where we would pre-allocate the space for the batch of data we would read in?